pyface/application.py

```diff
@@ -0,0 +1,259 @@
+            application = Application(id='hello-world', name='Hello World')
+            window = MainWindow()
+            application.on_trait_change(
+                lambda: application.add_window(window),
+                'application_initialized'
+            )
+            application.run()
+
+    """
+
+    # -------------------------------------------------------------------------
+    # 'Application' interface
+    # -------------------------------------------------------------------------
+
+    # Branding ----------------------------------------------------------------
+
+    #: The splash screen for the application. No splash screen by default
+    splash_screen = Instance(ISplashScreen)
+
+    #: Icon for the application
+    icon = Instance(IImageResource)
+
+    # Window management -------------------------------------------------------
+
+    #: Default window size
+    window_size = Tuple((800, 600))
+
+    #: Currently active Window if any
+    active_window = Instance(IWindow)
+
+    #: List of all open windows in the application
+    windows = List(Instance(IWindow))
+
+    #: The Pyface GUI instance for the application
+    gui = Instance(IGUI)
+
+    # Window lifecycle methods -----------------------------------------------
+
+    def add_window(self, window):
+        """ Add a new window to the windows we are tracking. """
+        if window.size == (-1, -1):
+            window.size = self.window_size
+        if not window.title:
+            window.title = self.name
+        if self.icon:
+            window.icon = self.icon
+
+        # Keep a handle on all windows created so that non-active windows don't
+        # get garbage collected
+        self.windows.append(window)
+
+        # set up window listeners
+        # NOTE: 'activated' is not fired on a window when the window first
+        # opens and gets focus. It is only fired when the window comes from
+        # lower in the stack to be the active window.
+        window.on_trait_change(self._on_activate_window, 'activated')
+        window.on_trait_change(self._on_window_closing, 'closing')
+        window.on_trait_change(self._on_window_closed, 'closed')
+        window.open()
+
+    # -------------------------------------------------------------------------
+    # 'BaseApplication' interface
+    # -------------------------------------------------------------------------
+
+    def start(self):
+        """ Start the application, setting up things that are required
+
+        Subclasses should open at least one ApplicationWindow or subclass in
+        their start method, and should call the superclass start() method
+        before doing any work themselves.
+        """
+        ok = super(Application, self).start()
+        if ok:
+            # set up logging so messages can be displayed on splash screen
+            self._setup_logging()
+
+            # create the GUI so that the splash screen comes up first thing
+            self.gui
```

Comment:
This code assumes that `self.gui` will trigger the default initializer which as a side effect will create a GUI will the desired splash screen. 

I would suggest the following

- There is no _gui_default ,method
- The gui trait is set only once here `self.gui = GUI(splash_screen=self.splash_screen)` and reset on stop.
- The gui trait is considered readonly